File: Leetcode/Python/_1020.py
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        queue = deque([])
        ROWS = len(grid)
        COLS = len(grid[0])
        total = 0
        for row in range(ROWS):
            for col in range(COLS):
                if grid[row][col]:
                    total += 1
                    if row == 0 or row == ROWS - 1 or col == 0 or col == COLS - 1:
                        queue.append((row, col))
                        grid[row][col] = 0

        while queue:
            r, c = queue.popleft()
            total -= 1
            for dx, dy in ((-1, 0), (0, -1), (1, 0), (0, 1)):
                row = r + dx
                col = c + dy
                if 0 <= row < ROWS and 0 <= col < COLS and grid[row][col] == 1:
                    grid[row][col] = 0
                    queue.append((row, col))

        return total
      
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        def dfs(row, col):
            if 0 < row < ROWS - 1 and 0 < col < COLS - 1 and grid[row][col] == 1:
                grid[row][col] = 3
                return 1 + dfs(row + 1, col) + dfs(row - 1, col) + dfs(row, col + 1) + dfs(row, col - 1)
            if 0 <= row <= ROWS - 1 and 0 <= col <= COLS - 1 and grid[row][col] == 0:
                return 0
            elif row == 0 or row == ROWS - 1 or col == 0 or col == COLS - 1 or grid[row][col] == 1:
                grid[row][col] = 2
                return -float("inf")
            return 0

        ROWS = len(grid)
        COLS = len(grid[0])
        cnt = 0
        for row in range(ROWS):
            for col in range(COLS):
                if grid[row][col] == 1:
                    out = dfs(row, col)
                    cnt += max(0, out)
        return cnt

File: Leetcode/cpp/1020.cpp
class Solution {
public:
    void dfs(int row, int col, int& ROWS, int& COLS, vector<vector<int>>& grid) {
      if (row < 0 || row > ROWS - 1 || col < 0 || col > COLS - 1) {
        return;
      } else if (grid[row][col] == 1) {
        grid[row][col] = 0;
        dfs(row + 1, col, ROWS, COLS, grid);
        dfs(row - 1, col, ROWS, COLS, grid);
        dfs(row, col + 1, ROWS, COLS, grid);
        dfs(row, col - 1, ROWS, COLS, grid);
        return;
      }
      return;
    }
    int numEnclaves(vector<vector<int>>& grid) {
      int ROWS = grid.size(), COLS = grid[0].size();
      for (int row = 0; row < ROWS; row++) {
        for (int col = 0; col < COLS; col++) {
          if ((row == 0 || col == 0 || row == ROWS - 1 || col == COLS - 1) &&grid[row][col] == 1) {
            dfs(row, col, ROWS, COLS, grid);
          }
        }
      }
      int count = 0;
      for (int row = 0; row < ROWS; row++) {
        for (int col = 0; col < COLS; col++) {
          if (grid[row][col]) {
              count++;
          }
        }
      }
      return count;
    }
};
